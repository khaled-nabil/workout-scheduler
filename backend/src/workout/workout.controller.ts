import {
    Controller,
    DefaultValuePipe,
    Get,
    Param,
    ParseIntPipe,
    Query,
} from '@nestjs/common';
import { ApiOperation, ApiResponse } from '@nestjs/swagger';
import { ObjectID } from 'typeorm';
import { Workout } from './workout.interface';
import { WorkoutService } from './workout.service';

@Controller('workout')
export class WorkoutController {
    constructor(private workoutService: WorkoutService) {}

    @ApiOperation({
        summary: 'Get paginated list of workouts',
        parameters: [
            { name: 'limit', in: 'query' },
            { name: 'offset', in: 'query' },
        ],
    })
    @ApiResponse({ status: 200, description: 'Return workouts.' })
    @ApiResponse({
        status: 404,
        description: 'No workouts found for query parameters.',
    })
    @Get()
    async getWorkouts(
        @Query('limit', new DefaultValuePipe(10), ParseIntPipe)
        limit: number,
        @Query('offset', new DefaultValuePipe(0), ParseIntPipe)
        offset: number
    ): Promise<Array<Workout>> {
        if (limit > 100) limit = 100;
        return await this.workoutService.findAll({ limit, offset });
    }
    @ApiOperation({
        summary: 'Get single workout by id',
        parameters: [{ name: 'id', in: 'path' }],
    })
    @ApiResponse({ status: 200, description: 'Return workout.' })
    @ApiResponse({
        status: 404,
        description: 'No workouk found with id.',
    })
    @Get(':id')
    async getWorkout(
        @Param('id')
        id: ObjectID
    ): Promise<Workout> {
        return await this.workoutService.findOne(id);
    }
}
