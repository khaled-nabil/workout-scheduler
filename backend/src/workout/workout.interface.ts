export interface Workout {
    name: string;
    description: string;
    startDate: Date;
    category: string;
}

export enum Category {
    c1 = 'c1',
    c2 = 'c2',
    c3 = 'c3',
    c4 = 'c4',
    c5 = 'c5',
    c6 = 'c6',
    c7 = 'c7',
}

// Service definitions
export interface FindAll {
    (pageniate: { limit: number; offset: number }): Promise<Array<Workout>>;
}
