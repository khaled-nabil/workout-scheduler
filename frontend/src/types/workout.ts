export interface Workout {
  _id: string;
  name: string;
  description: string;
  startDate: Date;
  category: string;
  picture: string;
}
