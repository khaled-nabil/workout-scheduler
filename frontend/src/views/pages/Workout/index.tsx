import React from 'react';
import { MainLayout } from 'views/templates';
import { LinkBox, Row } from 'views/molecules';
import useWorkout from './Workout.hook';
import { Box, Image, Paragraph } from 'views/atoms';
import moment from 'moment';

const Workout = () => {
  const { workout } = useWorkout();
  return (
    <MainLayout>
      <Row mx="auto" mb={3} width={[1, null, 3 / 4, 1 / 2]} px="4">
        <LinkBox href="/">{'< All Workouts'}</LinkBox>
      </Row>
      {workout && (
        <Row mx="auto" width={[1, null, 3 / 4, 1 / 2]} px="4">
          <Box width={[1, null, 1 / 4]}>
            <Image src={workout.picture} alt={workout.name} />
          </Box>
          <Box pl="4">
            <Paragraph>{workout.name}</Paragraph>
            <Paragraph>{workout.description}</Paragraph>
            <Paragraph>{workout.category}</Paragraph>
            <Paragraph>
              {moment(workout.startDate).format('YYYY-MM-DD')}
            </Paragraph>
          </Box>
        </Row>
      )}
    </MainLayout>
  );
};

export default Workout;
