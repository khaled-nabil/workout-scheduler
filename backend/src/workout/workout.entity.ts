import { Entity, Column, ObjectID, ObjectIdColumn } from 'typeorm';
import { Category, Workout } from './workout.interface';

@Entity('workout')
export class WorkoutEntity implements Workout {
    @ObjectIdColumn()
    id: ObjectID;

    @Column()
    name: string;

    @Column()
    description: string;

    @Column({ type: 'timestamp', default: () => 'CURRENT_TIMESTAMP' })
    startDate: Date;

    @Column('enum', { enum: Category })
    category: Category;

    @Column({ default: 'http://placehold.it/200x200' })
    picture: string;
}
