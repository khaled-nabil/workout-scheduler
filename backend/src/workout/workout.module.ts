import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { WorkoutController } from './workout.controller';
import { WorkoutEntity } from './workout.entity';
import { WorkoutService } from './workout.service';

@Module({
    imports: [TypeOrmModule.forFeature([WorkoutEntity])],
    controllers: [WorkoutController],
    providers: [WorkoutService],
})
export class WorkoutModule {}
