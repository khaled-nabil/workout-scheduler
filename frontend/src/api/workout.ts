import AxiosClient from './client';
import { Workout } from '../types/workout';

const getWorkouts = async (): Promise<Array<Workout>> => {
  const { data } = await AxiosClient.get<Array<Workout>>('/workout');

  return data;
};

export { getWorkouts };
