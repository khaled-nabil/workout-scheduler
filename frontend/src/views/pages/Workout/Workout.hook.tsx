import { useState, useMemo } from 'react';
import { getWorkout } from 'api/workout';
import { Workout } from 'types/workout';
import { useParams } from 'react-router-dom';

export type WorkoutsParams = { id: string };

const useWorkout = () => {
  const [workout, setWorkout] = useState<Workout>();
  const [loading, setLoading] = useState(true);
  const [hasError, setHasError] = useState(false);
  const { id } = useParams<WorkoutsParams>();

  useMemo(() => {
    const fetch = async () => {
      try {
        setLoading(true);
        setWorkout(await getWorkout({ id }));
        setLoading(false);
      } catch (error) {
        setHasError(true);
        setLoading(false);
      }
    };
    fetch();
  }, [id]);

  return { workout, loading, hasError };
};

export default useWorkout;
