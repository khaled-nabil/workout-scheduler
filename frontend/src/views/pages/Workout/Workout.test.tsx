import React from 'react';
import { shallow } from 'enzyme';
import { mocked } from 'ts-jest/utils';
import Workout from './';
import { Workout as WorkoutType } from 'types/workout';
import useWorkoutActual from './Workout.hook';
import toJson from 'enzyme-to-json';

jest.mock('./Workout.hook');

const useWorkout = mocked(useWorkoutActual, true);

const workoutData: WorkoutType = {
  _id: '123',
  name: 'elit Lorem amet',
  description: 'Eu laborum duis reprehenderit minim.',
  category: 'c1',
  picture: 'http://placehold.it/200x200',
  startDate: new Date(),
};

const useWorkoutResponse = (workout?: WorkoutType) => ({
  workout: workout,
  loading: false,
  hasError: false,
});

describe('Workout', () => {
  it('Should display workout details when workout date exists', () => {
    useWorkout.mockReturnValueOnce(useWorkoutResponse(workoutData));
    const component = shallow(<Workout />);
    const Image = component.find('Image');
    expect(Image).toHaveLength(1);
    expect(Image.props().src).toBe(workoutData.picture);
    const Paragraph = component.find('Paragraph');
    expect(Paragraph).toHaveLength(4);
    expect(toJson(component)).toMatchSnapshot();
  });
  it('Should display only back button when no workout passed', () => {
    useWorkout.mockReturnValueOnce(useWorkoutResponse());
    const component = shallow(<Workout />);
    const Image = component.find('Image');
    expect(Image.exists()).toBeFalsy();
    const Row = component.find('Row');
    expect(Row).toHaveLength(1);
    expect(toJson(component)).toMatchSnapshot();
  });
});
