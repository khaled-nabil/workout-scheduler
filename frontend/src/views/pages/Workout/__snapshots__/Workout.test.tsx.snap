// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workout Should display only back button when no workout passed 1`] = `
<MainLayout>
  <Row
    mb={3}
    mx="auto"
    px="4"
    width={
      Array [
        1,
        null,
        0.75,
        0.5,
      ]
    }
  >
    <Styled(Link)
      href="/"
    >
      &lt; All Workouts
    </Styled(Link)>
  </Row>
</MainLayout>
`;

exports[`Workout Should display workout details when workout date exists 1`] = `
<MainLayout>
  <Row
    mb={3}
    mx="auto"
    px="4"
    width={
      Array [
        1,
        null,
        0.75,
        0.5,
      ]
    }
  >
    <Styled(Link)
      href="/"
    >
      &lt; All Workouts
    </Styled(Link)>
  </Row>
  <Row
    mx="auto"
    px="4"
    width={
      Array [
        1,
        null,
        0.75,
        0.5,
      ]
    }
  >
    <Box
      width={
        Array [
          1,
          null,
          0.25,
        ]
      }
    >
      <Image
        alt="elit Lorem amet"
        src="http://placehold.it/200x200"
      />
    </Box>
    <Box
      pl="4"
    >
      <Paragraph>
        elit Lorem amet
      </Paragraph>
      <Paragraph>
        Eu laborum duis reprehenderit minim.
      </Paragraph>
      <Paragraph>
        c1
      </Paragraph>
      <Paragraph>
        2020-09-22
      </Paragraph>
    </Box>
  </Row>
</MainLayout>
`;
